akka-persistence/src/main/scala/akka/persistence/Eventsourced.scala

```diff
@@ -514,44 +523,56 @@ private[persistence] trait Eventsourced extends Snapshotter with PersistenceStas
    *
    * All incoming messages are stashed.
    */
-  private def recovering(recoveryBehavior: Receive, previousReceiveTimeout: Duration) =
+  private def recovering(recoveryBehavior: Receive) =
     new State {
+
+      // protect against snapshot stalling forever because of journal overloaded and such
+      val timeout = extension.journalConfigFor(journalPluginId).getMillisDuration("recovery-event-timeout")
+      val timeoutCancellable = {
+        import context.dispatcher
+        context.system.scheduler.schedule(timeout, timeout, self, RecoveryTick(snapshot = false))
+      }
+      var eventSeenInInterval = false
+
       override def toString: String = "replay started"
 
       override def recoveryRunning: Boolean = true
 
       override def stateReceive(receive: Receive, message: Any) = message match {
         case ReplayedMessage(p) ⇒
           try {
+            eventSeenInInterval = true
             updateLastSequenceNr(p)
             Eventsourced.super.aroundReceive(recoveryBehavior, p)
           } catch {
             case NonFatal(t) ⇒
+              timeoutCancellable.cancel()
               try onRecoveryFailure(t, Some(p.payload)) finally context.stop(self)
           }
         case RecoverySuccess(highestSeqNr) ⇒
-          resetRecieveTimeout()
+          timeoutCancellable.cancel()
```

Comment:
there is a small chance that a `RecoveryTick` is already enqueued in the mailbox here and will be delivered to user's `receive`. That could be filtered out inn`ProcessingState` or at least ignored in `unhandled`.